File: packages/datasets/src/contract.ts
import { sha256 } from '@noble/hashes/sha2';
import { bytesToHex } from '@noble/hashes/utils';
import {
  getDatasetCatalogs,
  type DatasetCatalog,
  type DatasetCatalogSource,
  type DimensionCatalogEntry,
  type MeasureCatalogEntry,
  type MetricCatalogEntry,
  type FilterCatalogEntry,
  type RelationshipCatalogEntry,
} from './catalog.js';
import type { DatasetLimits } from './types.js';

/**
 * Version of the semantic contract format. Bump when the serialized shape
 * changes in a way that snapshot consumers must account for.
 */
export const SEMANTIC_CONTRACT_VERSION = 1;

export interface ContractDimension {
  type: DimensionCatalogEntry['type'];
  column?: string;
  sql?: string;
  label?: string;
  description?: string;
  filterable: boolean;
  groupable: boolean;
}

export interface ContractMeasure {
  aggregation: MeasureCatalogEntry['aggregation'];
  field: string;
  sql?: string;
  label?: string;
  description?: string;
}

export interface ContractMetric {
  kind: MetricCatalogEntry['kind'];
  valueType: MetricCatalogEntry['valueType'];
  label?: string;
  description?: string;
  dimensions: string[];
  measures?: string[];
  filters: string[];
  grains: string[];
  grain?: string;
  requires?: string[];
}

export interface ContractFilter {
  field: string;
  label?: string;
  description?: string;
  operators: string[];
  valueType?: FilterCatalogEntry['valueType'];
}

export interface ContractRelationship {
  kind: RelationshipCatalogEntry['kind'];
  target: string;
  from: string;
  to: string;
}

export interface ContractDataset {
  name: string;
  source: string;
  tenantKey?: string;
  timeKey?: string;
  requiresTenant: boolean;
  supportedGrains: string[];
  dimensions: Record<string, ContractDimension>;
  measures: Record<string, ContractMeasure>;
  metrics: Record<string, ContractMetric>;
  filters: Record<string, ContractFilter>;
  relationships: Record<string, ContractRelationship>;
  limits?: DatasetLimits;
}

export interface SemanticContract {
  version: number;
  datasets: Record<string, ContractDataset>;
  contentHash: string;
}

type SemanticContractWithoutHash = Omit<SemanticContract, 'contentHash'>;

export interface SerializeSemanticContractOptions {
  /**
   * Include raw `sql` escape-hatch expressions for dimensions/measures.
   *
   * Defaults to `true` for trusted contexts (snapshots, CI, codegen) where the
   * SQL already lives in the developer's source. Set to `false` when serving the
   * contract to untrusted consumers so internal SQL is not exposed.
   */
  includeSql?: boolean;
}

/**
 * Builds a deterministic semantic contract from dataset instances.
 *
 * The contract is a normalized, sorted projection of the dataset catalog with a
 * version marker and content hash. Object keys and unordered arrays are sorted,
 * and SQL escape hatches are whitespace-normalized so logically equal models
 * produce identical JSON and hashes. This is the shared source consumed by
 * snapshots, diffs, CI validation, docs, and codegen.
 */
export function serializeSemanticContract(
  datasets: Record<string, DatasetCatalogSource>,
  options: SerializeSemanticContractOptions = {},
): SemanticContract {
  const includeSql = options.includeSql ?? true;
  const catalogs = getDatasetCatalogs(datasets);

  const contract: SemanticContractWithoutHash = {
    version: SEMANTIC_CONTRACT_VERSION,
    datasets: sortedRecord(
      Object.entries(catalogs).map(([name, catalog]) => [name, datasetToContract(catalog, includeSql)]),
    ),
  };

  return {
    ...contract,
    contentHash: hashContract(contract),
  };
}

function datasetToContract(catalog: DatasetCatalog, includeSql: boolean): ContractDataset {
  return {
    name: catalog.name,
    source: catalog.source,
    ...(catalog.tenantKey !== undefined ? { tenantKey: catalog.tenantKey } : {}),
    ...(catalog.timeKey !== undefined ? { timeKey: catalog.timeKey } : {}),
    requiresTenant: catalog.requiresTenant,
    supportedGrains: uniqueSorted(catalog.supportedGrains),
    dimensions: sortedRecord(
      Object.entries(catalog.dimensions).map(([name, entry]) => [name, dimensionToContract(entry, includeSql)]),
    ),
    measures: sortedRecord(
      Object.entries(catalog.measures).map(([name, entry]) => [name, measureToContract(entry, includeSql)]),
    ),
    metrics: sortedRecord(
      Object.entries(catalog.metrics).map(([name, entry]) => [name, metricToContract(entry)]),
    ),
    filters: sortedRecord(
      Object.entries(catalog.filters).map(([name, entry]) => [name, filterToContract(entry)]),
    ),
    relationships: sortedRecord(
      Object.entries(catalog.relationships).map(([name, entry]) => [name, relationshipToContract(entry)]),
    ),
    ...(catalog.limits !== undefined ? { limits: limitsToContract(catalog.limits) } : {}),
  };
}

/**
 * Emits dataset limits with a fixed key order. The source object's key order is
 * author-controlled, so normalizing it keeps the contract hash stable for
 * logically identical limits.
 */
function limitsToContract(limits: DatasetLimits): DatasetLimits {
  return {
    ...(limits.maxDimensions !== undefined ? { maxDimensions: limits.maxDimensions } : {}),
    ...(limits.maxFilters !== undefined ? { maxFilters: limits.maxFilters } : {}),
    ...(limits.maxMeasures !== undefined ? { maxMeasures: limits.maxMeasures } : {}),
    ...(limits.maxResultSize !== undefined ? { maxResultSize: limits.maxResultSize } : {}),
  };
}

function dimensionToContract(entry: DimensionCatalogEntry, includeSql: boolean): ContractDimension {
  return {
    type: entry.type,
    ...(entry.column !== undefined ? { column: entry.column } : {}),
    ...(includeSql && entry.sql !== undefined ? { sql: normalizeSql(entry.sql) } : {}),
    ...(entry.label !== undefined ? { label: entry.label } : {}),
    ...(entry.description !== undefined ? { description: entry.description } : {}),
    filterable: entry.filterable,
    groupable: entry.groupable,
  };
}

function measureToContract(entry: MeasureCatalogEntry, includeSql: boolean): ContractMeasure {
  return {
    aggregation: entry.aggregation,
    field: entry.field,
    ...(includeSql && entry.sql !== undefined ? { sql: normalizeSql(entry.sql) } : {}),
    ...(entry.label !== undefined ? { label: entry.label } : {}),
    ...(entry.description !== undefined ? { description: entry.description } : {}),
  };
}

function metricToContract(entry: MetricCatalogEntry): ContractMetric {
  return {
    kind: entry.kind,
    valueType: entry.valueType,
    ...(entry.label !== undefined ? { label: entry.label } : {}),
    ...(entry.description !== undefined ? { description: entry.description } : {}),
    dimensions: uniqueSorted(entry.dimensions),
    ...(entry.measures !== undefined ? { measures: uniqueSorted(entry.measures) } : {}),
    filters: uniqueSorted(entry.filters),
    grains: uniqueSorted(entry.grains),
    ...(entry.grain !== undefined ? { grain: entry.grain } : {}),
    ...(entry.requires !== undefined ? { requires: uniqueSorted(entry.requires) } : {}),
  };
}

function filterToContract(entry: FilterCatalogEntry): ContractFilter {
  return {
    field: entry.field,
    ...(entry.label !== undefined ? { label: entry.label } : {}),
    ...(entry.description !== undefined ? { description: entry.description } : {}),
    operators: uniqueSorted(entry.operators ?? []),
    ...(entry.valueType !== undefined ? { valueType: entry.valueType } : {}),
  };
}

function relationshipToContract(entry: RelationshipCatalogEntry): ContractRelationship {
  return {
    kind: entry.kind,
    target: entry.target,
    from: entry.from,
    to: entry.to,
  };
}

/**
 * Serializes a contract with stable formatting for writing to disk and hashing.
 */
export function contractToStableJson(
  contract: SemanticContract | SemanticContractWithoutHash,
): string {
  return JSON.stringify(contract, null, 2);
}

/**
 * Computes the SHA-256 content hash for a normalized contract.
 *
 * Uses `@noble/hashes` (audited, isomorphic, synchronous) rather than
 * `node:crypto` so the contract stays usable in browser/edge runtimes.
 */
export function hashContract(
  contract: SemanticContract | SemanticContractWithoutHash,
): string {
  return bytesToHex(sha256(contractToStableJson(contract)));
}

/**
 * Locale-independent string comparison by UTF-16 code unit. Used everywhere the
 * contract sorts, so the content hash is stable across environments (CI ICU
 * versions, locales) rather than depending on `localeCompare`.
 */
function compareStrings(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

/** Builds an object whose keys are inserted in sorted order for stable JSON. */
function sortedRecord<T>(entries: [string, T][]): Record<string, T> {
  return Object.fromEntries(
    [...entries].sort(([left], [right]) => compareStrings(left, right)),
  );
}

/** Deduplicates and sorts a list so logically-equal sets serialize identically. */
function uniqueSorted(values: readonly string[]): string[] {
  return Array.from(new Set(values)).sort(compareStrings);
}

/** Normalizes SQL escape-hatch whitespace so equivalent SQL hashes identically. */
function normalizeSql(sql: string): string {
  const lines = sql
    .replace(/\r\n/g, '\n')
    .split('\n')
    .map(line => line.replace(/\s+$/g, ''));

  while (lines.length > 0 && lines[0].trim() === '') {
    lines.shift();
  }
  while (lines.length > 0 && lines[lines.length - 1].trim() === '') {
    lines.pop();
  }

  const indents = lines
    .filter(line => line.trim().length > 0)
    .map(line => line.match(/^\s*/)?.[0].length ?? 0);
  const minIndent = indents.length > 0 ? Math.min(...indents) : 0;

  return lines
    .map(line => line.slice(minIndent))
    .join('\n')
    .trim();
}

File: python/hypequery/src/hypequery/datasets/contract.py
"""Deterministic semantic contract over a set of datasets.

The contract is a normalized, sorted projection of the dataset catalog with a
version marker and a content hash. Object keys and unordered arrays are sorted
and SQL escape hatches are whitespace-normalized, so two logically equal models
produce identical JSON and identical hashes. Snapshots, diffs, CI checks, docs,
and codegen all read this one shape.
"""

from __future__ import annotations

import hashlib
import re
from typing import Final, cast

from .catalog import (
    DatasetCatalog,
    DimensionCatalogEntry,
    FilterCatalogEntry,
    MeasureCatalogEntry,
    RelationshipCatalogEntry,
    get_dataset_catalogs,
)
from .registry import DatasetRegistry
from .utils.canonical_json import sorted_record, unique_sorted
from .utils.stable_json import stable_json

#: Version of the semantic contract format. Bump when the serialized shape
#: changes in a way snapshot consumers must account for.
SEMANTIC_CONTRACT_VERSION: Final = 3

_TRAILING_SPACE = re.compile(r"[^\S\n]+$", re.MULTILINE)
_LEADING_SPACE = re.compile(r"^[^\S\n]*")


def _number(value: float) -> float | int:
    """Emit an integral value as an integer, the way a JSON number reads.

    The reference implementation has one number type, so `1` never serializes
    as `1.0`. Python would, and that byte difference would change the hash.
    """

    return int(value) if float(value).is_integer() else value


def normalize_sql(sql: str) -> str:
    """Normalize escape-hatch whitespace so equivalent SQL hashes identically."""

    lines = _TRAILING_SPACE.sub("", sql.replace("\r\n", "\n")).split("\n")
    while lines and not lines[0].strip():
        lines.pop(0)
    while lines and not lines[-1].strip():
        lines.pop()
    indents = [
        len(cast(re.Match[str], _LEADING_SPACE.match(line)).group())
        for line in lines
        if line.strip()
    ]
    minimum = min(indents) if indents else 0
    return "\n".join(line[minimum:] for line in lines).strip()


def _dimension(entry: DimensionCatalogEntry, *, include_sql: bool) -> dict[str, object]:
    result: dict[str, object] = {"type": entry["type"]}
    if "column" in entry:
        result["column"] = entry["column"]
    if include_sql and "sql" in entry:
        result["sql"] = normalize_sql(entry["sql"])
    for key in ("label", "description"):
        if key in entry:
            result[key] = entry[key]
    result["filterable"] = entry["filterable"]
    result["groupable"] = entry["groupable"]
    return result


def _measure(entry: MeasureCatalogEntry, *, include_sql: bool) -> dict[str, object]:
    result: dict[str, object] = {
        "aggregation": entry["aggregation"],
        "field": entry["field"],
    }
    if "argField" in entry:
        result["argField"] = entry["argField"]
    if "level" in entry:
        result["level"] = _number(entry["level"])
    if include_sql and "sql" in entry:
        result["sql"] = normalize_sql(entry["sql"])
    for key in ("label", "description"):
        if key in entry:
            result[key] = entry[key]
    return result


def _filter(entry: FilterCatalogEntry) -> dict[str, object]:
    result: dict[str, object] = {"field": entry["field"]}
    for key in ("label", "description"):
        if key in entry:
            result[key] = entry[key]
    result["operators"] = unique_sorted(entry["operators"])
    if "valueType" in entry:
        result["valueType"] = entry["valueType"]
    return result


def _relationship(entry: RelationshipCatalogEntry) -> dict[str, object]:
    return {
        "kind": entry["kind"],
        "target": entry["target"],
        "from": entry["from"],
        "to": entry["to"],
        "queryable": entry["queryable"],
        "fields": unique_sorted(entry["fields"]),
    }


def _limits(limits: dict[str, int]) -> dict[str, object]:
    """Emit limits in a fixed key order.

    The authored order is not the contract's, so normalizing it keeps the hash
    stable for logically identical limits.
    """

    order = ("maxDimensions", "maxFilters", "maxMeasures", "maxResultSize")
    return {key: limits[key] for key in order if key in limits}


def _dataset(catalog: DatasetCatalog, *, include_sql: bool) -> dict[str, object]:
    result: dict[str, object] = {"name": catalog["name"], "source": catalog["source"]}
    for key in ("tenantKey", "timeKey"):
        if key in catalog:
            result[key] = catalog[key]
    result["requiresTenant"] = catalog["requiresTenant"]
    result["supportedGrains"] = unique_sorted(catalog["supportedGrains"])
    result["dimensions"] = sorted_record(
        {
            name: _dimension(entry, include_sql=include_sql)
            for name, entry in catalog["dimensions"].items()
        }
    )
    result["measures"] = sorted_record(
        {
            name: _measure(entry, include_sql=include_sql)
            for name, entry in catalog["measures"].items()
        }
    )
    # Python has no metric handles yet, so this is always empty; the key stays
    # so the serialized shape matches.
    result["metrics"] = sorted_record(catalog["metrics"])
    result["filters"] = sorted_record(
        {name: _filter(entry) for name, entry in catalog["filters"].items()}
    )
    result["relationships"] = sorted_record(
        {name: _relationship(entry) for name, entry in catalog["relationships"].items()}
    )
    if "limits" in catalog:
        result["limits"] = _limits(cast(dict[str, int], catalog["limits"]))
    return result


def serialize_semantic_contract(
    registry: DatasetRegistry, *, include_sql: bool = True
) -> dict[str, object]:
    """Build a deterministic semantic contract from a registry's datasets.

    `include_sql` defaults to True for trusted contexts — snapshots, CI,
    codegen — where the SQL already lives in the author's source. Pass False
    when serving the contract to untrusted consumers, so internal SQL never
    reaches them.
    """

    catalogs = get_dataset_catalogs(registry)
    contract: dict[str, object] = {
        "version": SEMANTIC_CONTRACT_VERSION,
        "datasets": sorted_record(
            {name: _dataset(catalog, include_sql=include_sql) for name, catalog in catalogs.items()}
        ),
    }
    return {**contract, "contentHash": hash_contract(contract)}


def contract_to_stable_json(contract: dict[str, object]) -> str:
    """Serialize a contract with stable formatting, for writing and hashing."""

    return stable_json(contract)


def hash_contract(contract: dict[str, object]) -> str:
    """Compute the SHA-256 content hash of a normalized contract."""

    return hashlib.sha256(contract_to_stable_json(contract).encode("utf-8")).hexdigest()
